.NET: document diagnostic logger (#3365)
getsentry/sentry-docs: src/platforms/dotnet/diagnostic-logger.mdx

---
title: Troubleshooting and diagnosing issues
description: "How to enable logging in Sentry SDK for troubleshooting purposes."
sidebar_order: 20
---

Sentry has its own logger that it uses to report diagnostics which may be useful when troubleshooting. To enable logging, set the `Debug` option to `true`:

```csharp

> SentrySdk.Init(o =>
> {
>     o.Dsn = "___PUBLIC_DSN___";
> 
>     // Enable debug mode to write diagnostic messages
>     o.Debug = true;
> });

Reviewer: What about skipping the SentrySdk. since that differs on ASP.NET Core, Xamarin etc.
```suggestion
options =>
{
    options.Dsn = "___PUBLIC_DSN___";

    // Enable debug mode to write diagnostic messages
    options.Debug = true;
});
```

Author: I would personally find that confusing, I think. I would that the snippets should probably be compilable.

Reviewer: You'd need to add an include of sorts and override. But it's hard to maintain.

If anyone went through setting up for their platform (i.e xamarin, aspnet) they know where to add the options.
I agree it's not ideal but being pragmatic, this is simpler. Right now we have many places with the wrong code snippet because assumes bare bones .NET

Reviewer: Otherwise please add the code snippet like: `dotnet.mdx` then `dotnet.xamarin.mdx` and `dotnet.aspnetcore.mdx`

```

By default, Sentry will write diagnostic messages to console. This may not be optimal in some circumstances, for example when running applications that don't have a visible console window attached.

> To override this behavior, you can provide your own implementation of `IDiagnosticLogger`. Here's an example implementation that appends diagnostic messages to a file:

Reviewer: ```suggestion
To override this behavior, you can provide your own implementation of `IDiagnosticLogger`. Here's an example implementation that naively appends diagnostic messages to a file:
```

using Sentry.Extensibility;
using Sentry.Protocol;

public class FileAppenderDiagnosticLogger : IDiagnosticLogger, IDisposable
{
    private readonly StreamWriter _writer;
    private readonly SentryLevel _minimalLevel;

    public FileAppenderDiagnosticLogger(string filePath, SentryLevel minimalLevel)
    {
        _writer = new StreamWriter(filePath);
        _minimalLevel = minimalLevel;
    }

    public bool IsEnabled(SentryLevel level) => level >= _minimalLevel;

    public void Log(SentryLevel logLevel, string message, Exception exception = null, params object[] args)
    {
        lock (_writer)
        {
            _writer.Write("{0} - {1}: ", DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH\\:mm\\:ss.ffffZ", DateTimeFormatInfo.InvariantInfo), logLevel);
            _writer.Write("");
            _writer.Write(message, args);
            if (exception is Exception)
            {
                _writer.Write("Exception: ");
                _writer.Write(exception);
            }
            _writer.WriteLine();
            _writer.Flush();
        }
    }

    public void Dispose()
    {
        lock (_writer)
        {
            _writer.Dispose();
        }
    }
}

> ```

Reviewer: What about pointing out this is just a "demo"

````suggestion
```

For a production environment you'd want to extend that so the file doesn't grow indefinitely. And are eventually rotated to avoid filling up the disk space
````

> To use your custom implementation of `IDiagnosticLogger`, pass it to the `DiagnosticLogger` option:

Reviewer: I suggest moving this block up, above the example logger.

And also, document that we ship this one already: https://github.com/getsentry/sentry-dotnet/blob/main/src/Sentry/Infrastructure/DebugDiagnosticLogger.cs

For apps such as WPF, WinForms, that's a better approach since users can see the log inside Visual Studio. Since there's no console.
```suggestion

# Change the diagnostic logger

To use your custom implementation of `IDiagnosticLogger`, pass it to the `DiagnosticLogger` option:
```

---

> title: Troubleshooting and diagnosing issues

Reviewer 2: ```suggestion
title: Diagnostic Logger
```

> description: "How to enable logging in Sentry SDK for troubleshooting purposes."

Reviewer 2: ```suggestion
description: "Learn more about enabling logging to help troubleshoot."
```

sidebar_order: 20
---

> Sentry has its own logger that it uses to report diagnostics which may be useful when troubleshooting. To enable logging, set the `Debug` option to `true`:

Reviewer 2: ```suggestion
Sentry has its own logger to report diagnostics that may be useful when troubleshooting. To enable logging, set the `Debug` option to `true`:
```

```csharp
SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";

    // Enable debug mode to write diagnostic messages
    o.Debug = true;
});
```

> # Change the diagnostic logger

Reviewer 2: ```suggestion
# Customize Implementation
```

> By default, Sentry will write diagnostic messages to console. This may not be optimal in some circumstances, for example when running applications that don't have a visible console window attached.

Reviewer 2: ```suggestion
By default, Sentry will write diagnostic messages to console. This may not be optimal in some circumstances; for example, when running applications that don't have a visible console window attached.
```

To use a custom implementation of `IDiagnosticLogger`, you can pass it to the `DiagnosticLogger` option. Sentry comes with two implementations out of the box: `ConsoleDiagnosticLogger` and `DebugDiagnosticLogger`.

```csharp
SentrySdk.Init(o =>
{
    o.Dsn = "___PUBLIC_DSN___";

    o.Debug = true;

    // Provide a custom logger
    o.DiagnosticLogger = new DebugDiagnosticLogger(SentryLevel.Debug);
});
```

> You can also create your own implementation of `IDiagnosticLogger` to fully customize the logging behavior. For example, here's an implementation that naively appends diagnostic messages to a file:

Reviewer 2: ```suggestion
You can also create your own implementation of `IDiagnosticLogger` to fully customize logging behavior. For example, to naively append diagnostic messages to a file:
```

using Sentry.Extensibility;
using Sentry.Protocol;

public class FileAppenderDiagnosticLogger : IDiagnosticLogger, IDisposable
{
    private readonly StreamWriter _writer;
    private readonly SentryLevel _minimalLevel;

    public FileAppenderDiagnosticLogger(string filePath, SentryLevel minimalLevel)
    {
        _writer = new StreamWriter(filePath);
        _minimalLevel = minimalLevel;
    }

    public bool IsEnabled(SentryLevel level) => level >= _minimalLevel;

    public void Log(SentryLevel logLevel, string message, Exception exception = null, params object[] args)
    {
        lock (_writer)
        {
            _writer.Write("{0} - {1}: ", DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH\\:mm\\:ss.ffffZ", DateTimeFormatInfo.InvariantInfo), logLevel);
            _writer.Write("");
            _writer.Write(message, args);
            if (exception is Exception)
            {
                _writer.Write("Exception: ");
                _writer.Write(exception);
            }
            _writer.WriteLine();
            _writer.Flush();
        }
    }

    public void Dispose()
    {
        lock (_writer)
        {
            _writer.Dispose();
        }
    }
}
```

> For a production environment you'd want to extend that so the file doesn't grow indefinitely and are eventually rotated to avoid filling up the disk space.

Reviewer 2: Move this up prior to the code sample, since readers often don't read to the end of the page, and the implementation detail provided here is important.

Author: @PeloWriter done. I also enclosed it in `<Note>` so that it doesn't break the reading flow, since the previous sentence ends in `:`.
